### ISSUE TITLE
Use primitive mode for memoizee

### ISSUE THREAD

**Issue by straker** [dequelabs/axe-core#5294]
Memoizee has a [primitive mode](https://github.com/medikoo/memoizee#primitive-mode) which performs much faster than the default mode. The one requirement is that the arguments being passed into the function must be stringifiable and produce unique strings. Most of our functions that we memoize take a virtual node, so if we add a `toString` function to virtual node that returns a unique string (such as the `nodeIndex`) then we could use primitive mode for most of our functions. Testing shows this improves functions like `isVisibleToScreenReaders` significantly by between 70%-90%.

We should update all uses of `memoize` to use primitive mode when passed a virtual node. Some functions do not take a virtual node and only take an HTML element. For these elements we can create a passthrough function that gets the virtual node from the node, then calls a memoized function that does the work. For example `getVisibleChildTextRects` could turn into:

```js
function getVisibleChildTextRects(node) {
  const vNode = getNodeFromTree(node);
  return getVisibleChildTextRectsVirtual(vNode);
}
const getVisibleChildTextRectsVirtual = memoize(
  function getVisibleChildTextRectsMemoized(vNode) {
    const node = vNode.actualNode;
   // rest of function
```

**Comment by straker** [comment]
Updating the function to have a passthrough for domNode to virtual node that we can memoize already has precedent in https://github.com/dequelabs/axe-core/blob/develop/lib/commons/dom/is-visible-to-screenreader.js

**Comment by straker** [comment]
Just a note, the linked pr only resolved half the this issue. The other half was about using primitive mode even for functions that take DOM nodes, which requires a pass through function. So the issue is still left open until we do that.

**Comment by chutchins25** [comment]
Picked up the passthrough half in #5324. Three of the five memoized DOM-node functions convert cleanly — `getVisibleChildTextRects` (your example), `toGrid`, and `getPseudoElementArea`. Two don't, and one of them is a design question for you rather than something I wanted to decide.

**`getScroll` — the passthrough approach conflicts with the `core/utils` boundary.** Adding one there fails our own eslint rule:

> Utils is meant for utility functions that work independently of axe's state; utilities that require the virtual tree to be set up should go in commons, not utils.

The rule has an `ignores` list, but it's described as files with "known uses of virtual node that are legacy before this rule was enforced", so adding `get-scroll.js` to it would be working against the rule rather than qualifying for it. As far as I can tell the options are:

1. Move `getScroll` to `commons` — it's public API as `axe.utils.getScroll`, so that's a bigger call.
2. Add it to the `ignores` list and accept the boundary erodes a little.
3. Leave it in default mode.

Same question applies to any other `core/utils` function we'd want to convert this way. Which would you prefer?

**`getCellPosition` — the second argument defeats primitive mode.** It takes `(cell, tableGrid)` where `tableGrid` is an array of cells. The primitive key would be `String(tableGrid)`, and every cell stringifies to `[object HTMLTableCellElement]`, so two structurally identical tables produce an identical key and the second one gets the first one's positions.

Keying on `cell` alone would be correct today — the position is a function of the cell, and all three callers pass `toGrid(findUp(cell, 'table'))` — but it makes correctness depend on no future caller passing a foreign grid. Left it in default mode; happy to do it with `{ length: 1 }` if you think that constraint is fine to rely on.

**On the numbers:** worth setting expectations — these three are performance-neutral. `getVisibleChildTextRects` doesn't register at all on a 4,000 paragraph page, and `toGrid` is 2ms on a 60-table page. Whole-run timing is flat in both directions across two rounds of 12. The benefit is consistency and removing the O(N²) lookup risk if they ever get hot, not a measurable win today.

One implementation note: `toGrid` keeps an uncached path, because it's called with tables outside the virtual tree — twelve table tests failed on `getNodeFromTree` returning `null` before I added the fallback.

**Comment by straker** [comment]
Closing as completed. The utils functions can't use virtual node as stated in @chutchins25 comment and we don't want to require virtual node support in utils, so they won't be able to make use of the virtual node passthrough. If we determine that these functions (mostly likely `getScroll`) are becoming a problem we can revisit later.
